src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/SourceGenJsonTypeInfoOfT.cs

```diff
@@ -64,23 +52,41 @@ public SourceGenJsonTypeInfo(JsonSerializerOptions options, JsonObjectInfoValues
             Func<JsonConverter<T>> converterCreator,
             object? createObjectWithArgs = null,
             object? addFunc = null)
-            : base(typeof(T), options)
+            : base(GetConverter(collectionInfo, converterCreator), options)
         {
-            ConverterStrategy strategy = collectionInfo.KeyInfo == null ? ConverterStrategy.Enumerable : ConverterStrategy.Dictionary;
-            JsonConverter<T> converter = new JsonMetadataServicesConverter<T>(converterCreator, strategy);
-
-            KeyType = converter.KeyType;
-            ElementType = converter.ElementType;
             KeyTypeInfo = collectionInfo.KeyInfo;
             ElementTypeInfo = collectionInfo.ElementInfo ?? throw new ArgumentNullException(nameof(collectionInfo.ElementInfo));
             NumberHandling = collectionInfo.NumberHandling;
-            PropertyInfoForTypeInfo = JsonMetadataServices.CreateJsonPropertyInfoForClassInfo(typeof(T), this, converter, options);
             SerializeHandler = collectionInfo.SerializeHandler;
             CreateObjectWithArgs = createObjectWithArgs;
             AddMethodDelegate = addFunc;
             SetCreateObjectFunc(collectionInfo.ObjectCreator);
         }
 
+        private static JsonConverter GetConverter(JsonObjectInfoValues<T> objectInfo)
```

Comment:
Nit: `CreateConverter` instead of `GetConverter`?

Reply:
we already call it GetConverter in other place so I matched the name here